### ISSUE TITLE
dataframe map_partitions doesn't resolve future arguments

### ISSUE THREAD

**Issue by gabrielelanaro** [dask/dask#3236]
In map partitions I'd like to use some large argument broadcasted in distributed memory. (dask 0.17.1) (distributed 1.21.0)

However, the future argument is not resolved.
```
import pandas as pd
import numpy as np
from dask.distributed import Client, LocalCluster
import dask.dataframe as ddf

cluster = LocalCluster()
client = Client(cluster)

df = pd.DataFrame(np.random.randint(0,100,size=(100, 4)), columns=list('ABCD'))
data = ddf.from_pandas(df, npartitions=4)
a_fut = client.scatter(1, broadcast=True)

def operate(df, a):
    # df is a pandas.DataFrame
    return df + a

data.map_partitions(operate, a_fut).compute()
```

Traceback:

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
~/anaconda3/envs/personalization.ranking.stats/lib/python3.6/site-packages/pandas/core/internals.py in eval(self, func, other, errors, try_cast, mgr)
   1317             values, values_mask, other, other_mask = self._try_coerce_args(
-> 1318                 transf(values), other)
   1319         except TypeError:

~/anaconda3/envs/personalization.ranking.stats/lib/python3.6/site-packages/pandas/core/internals.py in _try_coerce_args(self, values, other)
    700                 type(other).__name__,
--> 701                 type(self).__name__.lower().replace('Block', '')))
    702 

TypeError: cannot convert Future to an intblock

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
~/anaconda3/envs/personalization.ranking.stats/lib/python3.6/site-packages/pandas/core/ops.py in na_op(x, y)
   1201         try:
-> 1202             result = expressions.evaluate(op, str_rep, x, y, **eval_kwargs)
   1203         except TypeError:

~/anaconda3/envs/personalization.ranking.stats/lib/python3.6/site-packages/pandas/core/computation/expressions.py in evaluate(op, op_str, a, b, use_numexpr, **eval_kwargs)
    203     if use_numexpr:
--> 204         return _evaluate(op, op_str, a, b, **eval_kwargs)
    205     return _evaluate_standard(op, op_str, a, b)

~/anaconda3/envs/personalization.ranking.stats/lib/python3.6/site-packages/pandas/core/computation/expressions.py in _evaluate_standard(op, op_str, a, b, **eval_kwargs)
     63     with np.errstate(all='ignore'):
---> 64         return op(a, b)
     65 

TypeError: unsupported operand type(s) for +: 'int' and 'Future'

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
~/anaconda3/envs/personalization.ranking.stats/lib/python3.6/site-packages/pandas/core/internals.py in eval(self, func, other, errors, try_cast, mgr)
   1376             with np.errstate(all='ignore'):
-> 1377                 result = get_result(other)
   1378 

~/anaconda3/envs/personalization.ranking.stats/lib/python3.6/site-packages/pandas/core/internals.py in get_result(other)
   1345             else:
-> 1346                 result = func(values, other)
   1347 

~/anaconda3/envs/personalization.ranking.stats/lib/python3.6/site-packages/pandas/core/ops.py in na_op(x, y)
   1227                     with np.errstate(all='ignore'):
-> 1228                         result[mask] = op(xrav, y)
   1229             else:

TypeError: unsupported operand type(s) for +: 'int' and 'Future'

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
~/anaconda3/envs/personalization.ranking.stats/lib/python3.6/site-packages/dask/dataframe/utils.py in raise_on_meta_error(funcname)
    136     try:
--> 137         yield
    138     except Exception as e:

~/anaconda3/envs/personalization.ranking.stats/lib/python3.6/site-packages/dask/dataframe/core.py in _emulate(func, *args, **kwargs)
   3289     with raise_on_meta_error(funcname(func)):
-> 3290         return func(*_extract_meta(args, True), **_extract_meta(kwargs, True))
   3291 

<ipython-input-4-f39ca780a680> in operate(df, a)
     14     # df is a pandas.DataFrame
---> 15     return df + a
     16 

~/anaconda3/envs/personalization.ranking.stats/lib/python3.6/site-packages/pandas/core/ops.py in f(self, other, axis, level, fill_value)
   1266 
-> 1267             return self._combine_const(other, na_op)
   1268 

~/anaconda3/envs/personalization.ranking.stats/lib/python3.6/site-packages/pandas/core/frame.py in _combine_const(self, other, func, errors, try_cast)
   3986                                    errors=errors,
-> 3987                                    try_cast=try_cast)
   3988         return self._constructor(new_data)

~/anaconda3/envs/personalization.ranking.stats/lib/python3.6/site-packages/pandas/core/internals.py in eval(self, **kwargs)
   3434     def eval(self, **kwargs):
-> 3435         return self.apply('eval', **kwargs)
   3436 

~/anaconda3/envs/personalization.ranking.stats/lib/python3.6/site-packages/pandas/core/internals.py in apply(self, f, axes, filter, do_integrity_check, consolidate, **kwargs)
   3328             kwargs['mgr'] = self
-> 3329             applied = getattr(b, f)(**kwargs)
   3330             result_blocks = _extend_blocks(applied, result_blocks)

~/anaconda3/envs/personalization.ranking.stats/lib/python3.6/site-packages/pandas/core/internals.py in eval(self, func, other, errors, try_cast, mgr)
   1322                               errors=errors,
-> 1323                               try_cast=try_cast, mgr=mgr)
   1324 

~/anaconda3/envs/personalization.ranking.stats/lib/python3.6/site-packages/pandas/core/internals.py in eval(self, func, other, errors, try_cast, mgr)
   1383         except Exception as detail:
-> 1384             result = handle_error()
   1385 

~/anaconda3/envs/personalization.ranking.stats/lib/python3.6/site-packages/pandas/core/internals.py in handle_error()
   1366                 raise TypeError('Could not operate %s with block values %s' %
-> 1367                                 (repr(other), str(detail)))  # noqa
   1368             else:

TypeError: Could not operate <Future: status: finished, type: int, key: int-c0a8a20f903a4915b94db8de3ea63195> with block values unsupported operand type(s) for +: 'int' and 'Future'

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
<ipython-input-4-f39ca780a680> in <module>()
     15     return df + a
     16 
---> 17 data.map_partitions(operate, a_fut).compute()

~/anaconda3/envs/personalization.ranking.stats/lib/python3.6/site-packages/dask/dataframe/core.py in map_partitions(self, func, *args, **kwargs)
    541         >>> ddf.map_partitions(func).clear_divisions()  # doctest: +SKIP
    542         """
--> 543         return map_partitions(func, self, *args, **kwargs)
    544 
    545     @insert_meta_param_description(pad=12)

~/anaconda3/envs/personalization.ranking.stats/lib/python3.6/site-packages/dask/dataframe/core.py in map_partitions(func, *args, **kwargs)
   3322 
   3323     if meta is no_default:
-> 3324         meta = _emulate(func, *args, **kwargs)
   3325 
   3326     if all(isinstance(arg, Scalar) for arg in args):

~/anaconda3/envs/personalization.ranking.stats/lib/python3.6/site-packages/dask/dataframe/core.py in _emulate(func, *args, **kwargs)
   3288     """
   3289     with raise_on_meta_error(funcname(func)):
-> 3290         return func(*_extract_meta(args, True), **_extract_meta(kwargs, True))
   3291 
   3292 

~/anaconda3/envs/personalization.ranking.stats/lib/python3.6/contextlib.py in __exit__(self, type, value, traceback)
     97                 value = type()
     98             try:
---> 99                 self.gen.throw(type, value, traceback)
    100             except StopIteration as exc:
    101                 # Suppress StopIteration *unless* it's the same exception that

~/anaconda3/envs/personalization.ranking.stats/lib/python3.6/site-packages/dask/dataframe/utils.py in raise_on_meta_error(funcname)
    148                ).format(" in `{0}`".format(funcname) if funcname else "",
    149                         repr(e), tb)
--> 150         raise ValueError(msg)
    151 
    152 

ValueError: Metadata inference failed in `operate`.

Original error is below:
------------------------
TypeError("Could not operate <Future: status: finished, type: int, key: int-c0a8a20f903a4915b94db8de3ea63195> with block values unsupported operand type(s) for +: 'int' and 'Future'",)

Traceback:
---------
  File "/home/gabriele/anaconda3/envs/personalization.ranking.stats/lib/python3.6/site-packages/dask/dataframe/utils.py", line 137, in raise_on_meta_error
    yield
  File "/home/gabriele/anaconda3/envs/personalization.ranking.stats/lib/python3.6/site-packages/dask/dataframe/core.py", line 3290, in _emulate
    return func(*_extract_meta(args, True), **_extract_meta(kwargs, True))
  File "<ipython-input-4-f39ca780a680>", line 15, in operate
    return df + a
  File "/home/gabriele/anaconda3/envs/personalization.ranking.stats/lib/python3.6/site-packages/pandas/core/ops.py", line 1267, in f
    return self._combine_const(other, na_op)
  File "/home/gabriele/anaconda3/envs/personalization.ranking.stats/lib/python3.6/site-packages/pandas/core/frame.py", line 3987, in _combine_const
    try_cast=try_cast)
  File "/home/gabriele/anaconda3/envs/personalization.ranking.stats/lib/python3.6/site-packages/pandas/core/internals.py", line 3435, in eval
    return self.apply('eval', **kwargs)
  File "/home/gabriele/anaconda3/envs/personalization.ranking.stats/lib/python3.6/site-packages/pandas/core/internals.py", line 3329, in apply
    applied = getattr(b, f)(**kwargs)
  File "/home/gabriele/anaconda3/envs/personalization.ranking.stats/lib/python3.6/site-packages/pandas/core/internals.py", line 1323, in eval
    try_cast=try_cast, mgr=mgr)
  File "/home/gabriele/anaconda3/envs/personalization.ranking.stats/lib/python3.6/site-packages/pandas/core/internals.py", line 1384, in eval
    result = handle_error()
  File "/home/gabriele/anaconda3/envs/personalization.ranking.stats/lib/python3.6/site-packages/pandas/core/internals.py", line 1367, in handle_error
    (repr(other), str(detail)))  # noqa
```

**Comment by gabrielelanaro** [comment]
UPDATE: 

The problem is related to meta inference part.

```
data.map_partitions(operate, a_fut, meta={'A': int, 'B': int, 'C': int, "D": int}).compute()
```

**Comment by jcrist** [comment]
Fixed by #3256.
